src/sql/src/query_model/hir/hir_from_qgm.rs

```diff
@@ -24,23 +31,393 @@ impl From<Model> for HirRelationExpr {
+    /// `self.common_subgraphs`. Otherwise, it will be pushed to `self.converted`.
+    fn convert_subgraph(&mut self, model: &Model, box_id: &BoxId) {
+        let r#box = model.get_box(*box_id);
+        if self.common_subgraphs.get(&box_id).is_some() {
+            // Subgraph has already been converted.
+            return;
+        }
+        let hir = match &r#box.box_type {
+            BoxType::Get(Get { id, unique_keys }) => {
+                let typ = mz_repr::RelationType::new(
+                    r#box
+                        .columns
+                        .iter()
+                        .map(|c| {
+                            if let BoxScalarExpr::BaseColumn(BaseColumn { column_type, .. }) =
+                                &c.expr
+                            {
+                                column_type.clone()
+                            } else {
+                                // TODO: the validator should make this branch unreachable.
+                                panic!("expected all columns in Get BoxType to be BaseColumn");
+                            }
+                        })
+                        .collect::<Vec<_>>(),
+                )
+                .with_keys(unique_keys.clone());
+                HirRelationExpr::Get {
+                    id: mz_expr::Id::Global(*id),
+                    typ,
+                }
+            }
+            BoxType::Values(values) => HirRelationExpr::constant(
+                // Map `Vec<Vec<BoxScalarExpr>` to `Vec<Vec<Datum>>`
+                values
+                    .rows
+                    .iter()
+                    .map(|row| {
+                        row.iter()
+                            .map(|expr| {
+                                if let BoxScalarExpr::Literal(datum, _) = expr {
+                                    datum.unpack_first()
+                                } else {
+                                    // TODO: the validator should make this branch unreachable.
+                                    panic!("expected all scalars in Values BoxType to be literals");
+                                }
+                            })
+                            .collect_vec()
+                    })
+                    .collect_vec(),
+                mz_repr::RelationType::new(r#box.attributes.get::<BoxRelationType>().to_owned()),
+            ),
+            BoxType::Select(select) => {
+                let (rels, scalars) = self.convert_quantifiers(&r#box);
+                let (join_q_ids, mut join_inputs): (Vec<_>, Vec<_>) = rels.into_iter().unzip();
+                let (map_q_ids, maps): (Vec<_>, Vec<_>) = scalars.into_iter().unzip();
+
+                let mut result = if let Some(first) = join_inputs.pop() {
+                    first
+                } else {
+                    HirRelationExpr::constant(vec![vec![]], mz_repr::RelationType::new(vec![]))
```

Comment:
I actually lean in the opposite direction, namely removing the restriction that a `Select` box has at least one non-`PreservedForeach` quantifier. 

I think that a `Select` box with zero or one quantifiers makes sense. If we interpret a `Select` box as a cross join + filters + projection, then:
* a `Select` box with no quantifiers can be interpreted as `<join identity>` + filters + projection.
* a `Select` box with one quantifier as (`<join identity>` cross join `<result of the quantifier>`) + filters + projection.

(`<join identity>`, of course, is one row with zero columns.)

I think that it would save space and inspection effort to simply delete from the Query Graph Model every join identity, i.e. every quantifier in a select box that points at a values box with one row and zero columns. Since a join identity has zero columns, no projection or predicate would ever refer to it. The presence of the join identity also not affect any rewrite or attribute calculation.

Reply:
Parking this discussion for later in #11375.